runtime/src/unchecked_extrinsic.rs

```diff
@@ -0,0 +1,437 @@
+	type SignedExtensions = Extra;
+}
+
+impl<Address, Call, Extra> Decode for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Decode,
+	Call: Decode,
+	Extra: SignedExtension,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		// This is a little more complicated than usual since the binary format must be compatible
+		// with SCALE's generic `Vec<u8>` type. Basically this just means accepting that there
+		// will be a prefix of vector length.
+		let expected_length: Compact<u32> = Decode::decode(input)?;
+		let before_length = input.remaining_len()?;
+
+		let version = input.read_byte()?;
+
+		let is_signed = version & 0b1000_0000 != 0;
+		let version = version & 0b0111_1111;
+		if version != EXTRINSIC_FORMAT_VERSION {
+			return Err("Invalid transaction version".into())
+		}
+
+		let signature = is_signed.then(|| Decode::decode(input)).transpose()?;
+		let function = Decode::decode(input)?;
+
+		if let Some((before_length, after_length)) =
+			input.remaining_len()?.and_then(|a| before_length.map(|b| (b, a)))
+		{
+			let length = before_length.saturating_sub(after_length);
+
+			if length != expected_length.0 as usize {
+				return Err("Invalid length prefix".into())
+			}
+		}
+
+		Ok(Self { signature, function })
+	}
+}
+
+impl<Address, Call, Extra> Encode for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Encode,
+	Call: Encode,
+	Extra: SignedExtension,
+{
+	fn encode(&self) -> Vec<u8> {
+		let mut tmp = Vec::with_capacity(sp_std::mem::size_of::<Self>());
+
+		// 1 byte version id.
+		match self.signature.as_ref() {
+			Some(s) => {
+				tmp.push(EXTRINSIC_FORMAT_VERSION | 0b1000_0000);
+				s.encode_to(&mut tmp);
+			},
+			None => {
+				tmp.push(EXTRINSIC_FORMAT_VERSION & 0b0111_1111);
+			},
+		}
+		self.function.encode_to(&mut tmp);
+
+		let compact_len = codec::Compact::<u32>(tmp.len() as u32);
+
+		// Allocate the output buffer with the correct length
+		let mut output = Vec::with_capacity(compact_len.size_hint() + tmp.len());
+
+		compact_len.encode_to(&mut output);
+		output.extend(tmp);
+
+		output
+	}
+}
+
+impl<Address, Call, Extra> EncodeLike for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Encode,
+	Call: Encode,
+	Extra: SignedExtension,
+{
+}
+
+#[cfg(feature = "std")]
+impl<Address: Encode, Call: Encode, Extra: SignedExtension> serde::Serialize
+	for UncheckedExtrinsic<Address, Call, Extra>
+{
+	fn serialize<S>(&self, seq: S) -> Result<S::Ok, S::Error>
+	where
+		S: ::serde::Serializer,
+	{
+		self.using_encoded(|bytes| seq.serialize_bytes(bytes))
+	}
+}
+
+#[cfg(feature = "std")]
+impl<'a, Address: Decode, Call: Decode, Extra: SignedExtension> serde::Deserialize<'a>
+	for UncheckedExtrinsic<Address, Call, Extra>
+{
+	fn deserialize<D>(de: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'a>,
+	{
+		let r = sp_core::bytes::deserialize(de)?;
+		Decode::decode(&mut &r[..])
+			.map_err(|e| serde::de::Error::custom(format!("Decode error: {e}")))
```

Comment:
If we use 'custom(e.to_string())' we can drop extra allocation from 'format!'

Reply:
yup, @nakul1010